lib/src/generator_frontend.dart

```diff
@@ -0,0 +1,377 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io' show File;
+
+import 'package:dartdoc/src/generator.dart';
+import 'package:dartdoc/src/logging.dart';
+import 'package:dartdoc/src/model/model.dart';
+import 'package:dartdoc/src/model_utils.dart';
+import 'package:dartdoc/src/warnings.dart';
+import 'package:path/path.dart' as path;
+
+typedef FileWriter = File Function(String filePath, Object content,
+    {bool allowOverwrite, Warnable element});
+
+/// [Generator] that delegates rendering to a [GeneratorBackend] and delegates
+/// file creation to a [FileWriter].
+class GeneratorFrontEnd implements Generator {
+  final GeneratorBackend _generatorBackend;
+  final FileWriter _writer;
+
+  // Used in write(). This being not null signals the generator is active.
+  String _outputDirectory;
+
+  final StreamController<void> _onFileCreated = StreamController(sync: true);
+
+  @override
+  Stream<void> get onFileCreated => _onFileCreated.stream;
+
+  @override
+  final Map<String, Warnable> writtenFiles = {};
+
+  GeneratorFrontEnd(this._generatorBackend, this._writer);
+
+  // Implement FileWriter so we can wrap the one given to us.
```

Comment:
This part feels a little weird, but we need to accomplish two things:

- Fix path separators
- Apply the `outputDirectory` given to us, required by signature of `generate()`

I think delegating to a writer is still appropriate because it can be mocked in tests. Perhaps with the writer being a separate component, we could remove `outputDirectory` from the signature of `Generator.generate()` and instead have `DartdocFileWriter` do this work instead.

Reply:
There is an awful lot of joining/splitting of outputDirectory going on (and that was in the original).  Anything we can do to reduce the number of times we are recombining paths seems like a good idea; I've just never went through and cleaned this up.

Reply:
Latest push no longer has this "wrapping" structure, and all this path manipulation happens in one place, `DartdocFileWriter`.